mlflow/utils/virtualenv.py

```diff
@@ -237,15 +238,32 @@ def _create_virtualenv(
 
     _logger.info("Installing dependencies")
     for deps in filter(None, [python_env.build_dependencies, python_env.dependencies]):
-        # Create a temporary requirements file in the model directory to resolve the references
-        # in it correctly.
-        tmp_req_file = f"requirements.{uuid.uuid4().hex}.txt"
-        local_model_path.joinpath(tmp_req_file).write_text("\n".join(deps))
-        try:
+        with TempDir() as t:
+            # Create a temporary requirements file in the model directory to resolve the references
+            # in it correctly. To do this, we must first symlink or copy the model directory's
+            # contents to a temporary location for compatibility with deployment tools that store
+            # models in a read-only mount
+            model_dir = t.path("model")
+            try:
+                os.makedirs(model_dir)
+                for model_item in os.listdir(local_model_path):
+                    os.symlink(
+                        src=os.path.join(local_model_path, model_item),
+                        dst=os.path.join(model_dir, model_item),
+                    )
+            except Exception as e:
+                _logger.warning(
+                    "Failed to symlink model directory during dependency installation"
+                    " Copying instead. Exception: %s",
+                    e,
+                )
+                shutil.rmtree(model_dir)
+                shutil.copytree(local_model_path, model_dir)
+
+            tmp_req_file = f"requirements.{uuid.uuid4().hex}.txt"
+            Path(model_dir).joinpath(tmp_req_file).write_text("\n".join(deps))
             cmd = _join_commands(activate_cmd, f"python -m pip install --quiet -r {tmp_req_file}")
-            _exec_cmd(cmd, capture_output=capture_output, cwd=local_model_path, extra_env=extra_env)
-        finally:
-            local_model_path.joinpath(tmp_req_file).unlink()
+            _exec_cmd(cmd, capture_output=capture_output, cwd=model_dir, extra_env=extra_env)
```

Comment:
does it matter if there are hanging symlinks? Would they need to have `os.unlink()` called on the destination path or is that irrelevant?

Reply:
`os.unlink()` is the same as `os.remove()`. The key point here is that the `TempDir` context manager removes the temp directory upon context manager exit, and remove / unlink deletes symlinks too.